File: QTS/3° Bim/app-para-teste/web/app/models/Emprestimo.php
<?php
// Model Emprestimo
// ATENÇÃO: código intencionalmente com falhas para fins de estudo.

require_once __DIR__ . '/../core/Database.php';

class Emprestimo
{
    private $db;

    // Limite de livros por aluno (constante mal utilizada - ver bugs no emprestar())
    const LIMITE_LIVROS = 3;

    public function __construct()
    {
        $this->db = Database::getInstance();
    }

    public function todos()
    {
        $sql = "SELECT e.*, a.nome AS aluno_nome, l.titulo AS livro_titulo
                FROM emprestimos e
                JOIN alunos a ON a.id = e.aluno_id
                JOIN livros l ON l.id = e.livro_id
                ORDER BY e.data_emprestimo DESC";
        $resultado = $this->db->query($sql);
        if ($resultado === false) {
            return [];
        }
        return $resultado->fetchAll(PDO::FETCH_ASSOC);
    }

    // Registra um novo empréstimo
    public function emprestar($alunoId, $livroId)
    {
        $livro = new Livro();
        $aluno = new Aluno();
        $livroExiste = $livro->porId($livroId);
        $alunoExiste = $aluno->buscar($alunoId);

        if (!$livroExiste) {
            return ['erro' => 'Livro não encontrado.'];
        }
        if (!$alunoExiste) {
            return ['erro' => 'Aluno não encontrado.'];
        }

        // BUG LÓGICO (regra de negócio): deveria impedir o empréstimo quando
        // a quantidade disponível é 0, mas a variável $disponivel é calculada
        // e NUNCA checada antes de inserir.
        $disponivel = $livro->quantidadeDisponivel($livroId);

        // BUG DE NEGÓCIO: limite de livros por aluno é checado contra o LIMITE
        // com uma condição invertida (>= em vez de >), permitindo empréstimos a mais.
        $sql = "SELECT COUNT(*) AS t FROM emprestimos
                WHERE aluno_id = $alunoId AND status = 'emprestado'";
        $r = $this->db->query($sql);
        $emprestados = (int)$r->fetch(PDO::FETCH_ASSOC)['t'];

        if ($emprestados >= self::LIMITE_LIVROS) {
            return ['erro' => 'Aluno já atingiu o limite de empréstimos.'];
        }

        // BUG LÓGICO: insere mesmo se não houver disponibilidade
        $data = date('Y-m-d');
        $sql = "INSERT INTO emprestimos (aluno_id, livro_id, data_emprestimo, status)
                VALUES ($alunoId, $livroId, '$data', 'emprestado')";
        $this->db->exec($sql);
        return ['id' => $this->db->getPdo()->lastInsertId()];
    }

    // Registra devolução de um empréstimo
    public function devolver($id)
    {
        // BUG: não verifica se o empréstimo já foi devolvido -> pode "devolver" duas vezes
        $data = date('Y-m-d');
        $sql = "UPDATE emprestimos
                SET data_devolucao = '$data', status = 'devolvido'
                WHERE id = $id";
        return $this->db->exec($sql);
    }

    // Lista empréstimos em atraso (data_devolucao nula e mais de 14 dias)
    public function atrasados()
    {
        $sql = "SELECT e.*, a.nome AS aluno_nome, l.titulo AS livro_titulo
                FROM emprestimos e
                JOIN alunos a ON a.id = e.aluno_id
                JOIN livros l ON l.id = e.livro_id
                WHERE e.status = 'emprestado'
                  AND date(e.data_emprestimo) < date('now', '-14 day')";
        $resultado = $this->db->query($sql);
        if ($resultado === false) {
            return [];
        }
        return $resultado->fetchAll(PDO::FETCH_ASSOC);
    }

    public function doAluno($alunoId)
    {
        $sql = "SELECT e.*, l.titulo AS livro_titulo
                FROM emprestimos e
                JOIN livros l ON l.id = e.livro_id
                WHERE e.aluno_id = $alunoId";
        $resultado = $this->db->query($sql);
        if ($resultado === false) {
            return [];
        }
        return $resultado->fetchAll(PDO::FETCH_ASSOC);
    }
}

File: QTS/3° Bim/app-para-teste/desktop/src/main/java/br/edu/etec/biblioteca/models/Emprestimo.java
package br.edu.etec.biblioteca.models;

import br.edu.etec.biblioteca.core.Database;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

/**
 * Model Emprestimo.
 * ATENÇÃO: código intencionalmente com FALHAS (regras de negócio quebradas).
 */
public class Emprestimo {

    // Limite de livros por aluno (constante no código).
    private static final int LIMITE_LIVROS = 3;

    private final Connection conn;

    public Emprestimo() {
        this.conn = Database.getConnection();
    }

    public List<String[]> todos() {
        List<String[]> lista = new ArrayList<>();
        String sql = "SELECT e.id, a.nome AS aluno_nome, l.titulo AS livro_titulo, "
            + "e.data_emprestimo, e.data_devolucao, e.status "
            + "FROM emprestimos e "
            + "JOIN alunos a ON a.id = e.aluno_id "
            + "JOIN livros l ON l.id = e.livro_id "
            + "ORDER BY e.data_emprestimo DESC";
        try (Statement st = conn.createStatement();
             ResultSet rs = st.executeQuery(sql)) {
            while (rs.next()) {
                lista.add(new String[]{
                    String.valueOf(rs.getInt("id")),
                    rs.getString("aluno_nome"),
                    rs.getString("livro_titulo"),
                    rs.getString("data_emprestimo"),
                    rs.getString("data_devolucao") == null ? "" : rs.getString("data_devolucao"),
                    rs.getString("status")
                });
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return lista;
    }

    /**
     * Registra um empréstimo.
     * BUG DE NEGÓCIO (crítico): a disponibilidade é calculada mas NUNCA checada
     * antes de inserir — é possível emprestar além do estoque.
     * BUG DE NEGÓCIO: o limite de livros por aluno é checado ANTES do insert e
     * usa ">= LIMITE", o que permite ao aluno pegar o 4º livro (ver docs).
     */
    public boolean emprestar(int alunoId, int livroId) {
        Livro livro = new Livro();
        Aluno aluno = new Aluno();

        if (livro.porId(livroId) == null) {
            System.out.println("[BUG] livro não encontrado (ignorado pela UI)");
            return false;
        }
        if (aluno.buscar(alunoId) == null) {
            System.out.println("[BUG] aluno não encontrado (ignorado pela UI)");
            return false;
        }

        int disponivel = livro.quantidadeDisponivel(livroId);
        // NOTA: variável calculada mas NUNCA usada -> o empréstimo segue mesmo com 0.

        // Contagem de empréstimos ativos do aluno
        int emprestados = 0;
        String countSql = "SELECT COUNT(*) AS t FROM emprestimos "
            + "WHERE aluno_id = " + alunoId + " AND status = 'emprestado'";
        try (Statement st = conn.createStatement();
             ResultSet rs = st.executeQuery(countSql)) {
            rs.next();
            emprestados = rs.getInt("t");
        } catch (SQLException e) {
            e.printStackTrace();
        }

        if (emprestados >= LIMITE_LIVROS) {
            System.out.println("[INFO] aluno já atingiu o limite (" + LIMITE_LIVROS + ")");
            return false;
        }

        String data = LocalDate.now().toString();
        String sql = "INSERT INTO emprestimos (aluno_id, livro_id, data_emprestimo, status) VALUES ("
            + alunoId + ", " + livroId + ", '" + data + "', 'emprestado')";
        try (Statement st = conn.createStatement()) {
            st.executeUpdate(sql);
            return true;
        } catch (SQLException e) {
            return false;
        }
    }

    /**
     * Devolve um empréstimo.
     * BUG DE NEGÓCIO: não verifica se o empréstimo já foi devolvido, permitindo
     * devolução dupla que sobrescreve a data.
     */
    public boolean devolver(int id) {
        String data = LocalDate.now().toString();
        String sql = "UPDATE emprestimos SET data_devolucao = '" + data
            + "', status = 'devolvido' WHERE id = " + id;
        try (Statement st = conn.createStatement()) {
            st.executeUpdate(sql);
            return true;
        } catch (SQLException e) {
            return false;
        }
    }

    /** Empréstimos ativos com mais de 14 dias (em atraso). */
    public List<String[]> atrasados() {
        List<String[]> lista = new ArrayList<>();
        String sql = "SELECT e.id, a.nome AS aluno_nome, l.titulo AS livro_titulo, e.data_emprestimo "
            + "FROM emprestimos e "
            + "JOIN alunos a ON a.id = e.aluno_id "
            + "JOIN livros l ON l.id = e.livro_id "
            + "WHERE e.status = 'emprestado' "
            + "  AND date(e.data_emprestimo) < date('now','-14 day')";
        try (Statement st = conn.createStatement();
             ResultSet rs = st.executeQuery(sql)) {
            while (rs.next()) {
                lista.add(new String[]{
                    rs.getString("aluno_nome"),
                    rs.getString("livro_titulo"),
                    rs.getString("data_emprestimo")
                });
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return lista;
    }
}
